[CI] CoordinatorTests testElectionSchedulingAfterDiscoveryOutage failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #3484 / amazonlinux-2_platform-support-aws](https://gradle-enterprise.elastic.co/s/da7jgsccpuiji)
- [elasticsearch-periodic-platform-support #3424 / windows-2016_checkpart1_platform-support-windows](https://gradle-enterprise.elastic.co/s/7bd7lnux5mplo)

**Reproduction Line:**
```
./gradlew ":server:test" --tests "org.elasticsearch.cluster.coordination.CoordinatorTests.testElectionSchedulingAfterDiscoveryOutage" -Dtests.seed=DC077D013F80A90B -Dtests.locale=sr-CS -Dtests.timezone=Africa/Harare -Druntime.java=22
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testElectionSchedulingAfterDiscoveryOutage'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.cluster.coordination.CoordinatorTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure Message:**
```
java.lang.AssertionError: node2 has applied its state 
Expected: <19L>
     but: was <18L>
```

**Issue Reasons:**
- [main] 2 failures in test testElectionSchedulingAfterDiscoveryOutage (0.2% fail rate in 892 executions)
- [main] 2 failures in pipeline elasticsearch-periodic-platform-support (12.5% fail rate in 16 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java
+++ b/server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java
@@ -230,9 +230,17 @@ public class PublicationTransportHandler {
     private void acceptState(ClusterState incomingState, ActionListener<PublishWithJoinResponse> actionListener) {
         assert incomingState.nodes().isLocalNodeElectedMaster() == false
             : "should handle local publications locally, but got " + incomingState;
-        clusterCoordinationExecutor.execute(
-            ActionRunnable.supply(actionListener, () -> handlePublishRequest.apply(new PublishRequest(incomingState)))
-        );
+        clusterCoordinationExecutor.execute(ActionRunnable.supply(actionListener, new CheckedSupplier<>() {
+            @Override
+            public PublishWithJoinResponse get() {
+                return handlePublishRequest.apply(new PublishRequest(incomingState));
+            }
+
+            @Override
+            public String toString() {
+                return "acceptState[term=" + incomingState.term() + ",version=" + incomingState.version() + "]";
+            }
+        }));
     }
 
     public PublicationContext newPublicationContext(ClusterStatePublicationEvent clusterStatePublicationEvent) {

--- a/server/src/main/java/org/elasticsearch/discovery/PeerFinder.java
+++ b/server/src/main/java/org/elasticsearch/discovery/PeerFinder.java
@@ -496,6 +496,11 @@ public abstract class PeerFinder {
                             } // else this Peer has been superseded by a different instance which should be left in place
                         }
                     }
+
+                    @Override
+                    public String toString() {
+                        return "Peer#establishConnection[" + transportAddress + "]";
+                    }
                 })
             );
         }
@@ -584,6 +589,7 @@ public abstract class PeerFinder {
                 + Optional.ofNullable(probeConnectionResult.get())
                     .map(result -> result.getDiscoveryNode().descriptionWithoutAttributes())
                     .orElse("unknown")
+                + "]"
                 + (peersRequestInFlight ? " [request in flight]" : "");
         }
     }

--- a/server/src/main/java/org/elasticsearch/transport/TransportService.java
+++ b/server/src/main/java/org/elasticsearch/transport/TransportService.java
@@ -1396,6 +1396,11 @@ public class TransportService extends AbstractLifecycleComponent
                     protected void doRun() {
                         handler.handleException(exception);
                     }
+
+                    @Override
+                    public String toString() {
+                        return "onConnectionClosed/handleException[" + handler + "]";
+                    }
                 });
             }
         }

--- a/server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java
+++ b/server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java
@@ -2047,7 +2047,15 @@ public class CoordinatorTests extends AbstractCoordinatorTestCase {
                     + 4 * DEFAULT_DELAY_VARIABILITY
                     // Then a commit of the new leader's first cluster state
                     + DEFAULT_CLUSTER_STATE_UPDATE_DELAY
-                    // Then the remaining node may join
+                    // Then the remaining node may experience a disconnect (see comment in PeerFinder#closePeers) for which it is
+                    // removed from the cluster, and its fault detection must also detect its removal
+                    + Math.max(
+                        DEFAULT_CLUSTER_STATE_UPDATE_DELAY,
+                        defaultMillis(LEADER_CHECK_TIMEOUT_SETTING) * LEADER_CHECK_RETRY_COUNT_SETTING.get(Settings.EMPTY)
+                    )
+                    // then it does another round of discovery
+                    + defaultMillis(DISCOVERY_FIND_PEERS_INTERVAL_SETTING)
+                    // and finally it joins the master
                     + DEFAULT_CLUSTER_STATE_UPDATE_DELAY
             );
 

